kiosk/src/Components/AddingGame.tsx

```diff
@@ -58,60 +61,82 @@ const AddingGame: React.FC<IProps> = ({ kiosk }) => {
+        const whenToPoll = 5000; // wait for 5 seconds to poll for game code data
+        if (kiosk.state !== KioskState.AddingGame) {
+            return;
+        }
+        return new Promise<void>(async (resolve, reject) => {
+            const getGameCode = async () => {
+                try {
+                    const gameCode: string = await getGameCodeAsync(kioskCode);
+                    await kiosk.saveNewGameAsync(gameCode);
+                    clearTimeout(pollFrequency);
+                    clearTimeout(pollTimeout);
+                    resolve();
+                    kiosk.launchGame(gameCode);
+                }
+                catch (error: any) {
+                    if (error.message === "Not Found" || 
+                        error.message === "Too Many Requests" ||
+                        error.message === "Failed to fetch") {
+                            clearTimeout(pollTimeout);
+                            clearTimeout(pollFrequency);
+                            reject();
+                            return;
+                    }
+                    pollFrequency = setTimeout(async () => {
+                        await getGameCode();
+                    }, whenToPoll)
+                }
+            };
+
+            pollTimeout = setTimeout(() => {
+                clearTimeout(pollTimeout);
+                clearTimeout(pollFrequency);
+                reject();
+                setRenderQRCode(false);
+                pollingStarted.current = false;
+            }, timeoutDuration)
+
+            await getGameCode();
+        });
+    }
+
     useEffect(() => {
-        async function generateKioskCode() {
+        const generateKioskCode = async () => {
             try {
                 const newKioskCode: string = await generateKioskCodeAsync();
                 setKioskCode(newKioskCode);
                 try {
-                    await addGameToKiosk(newKioskCode);
+                    await pollForGames(newKioskCode);
                 } catch (error) {
-                    console.log("Unable to add game to kioksk");
+                    console.log("the promise rejected");
+                    // TODO: dispatch error for time out here
                 }
             }
             catch (error) {
-                setRenderQRCode(false);
                 throw new Error("Unable to generate kiosk code");
+                //TODO: too many set here
             }
         }
 
-        generateKioskCode();
+        if (pollingStarted.current && renderQRCode) {
+            generateKioskCode();
+        }
 
-        function addGameToKiosk(kioskCode: string) {
-            const timeoutDuration = 600000; // wait for 10 minutes until the kiosk code expires
-            const whenToPoll = 5000; // wait for 5 seconds to poll for game code data
-            if (kiosk.state !== KioskState.AddingGame) {
-                return;
+        return () => {
+            pollingStarted.current = true;
```

Comment:
`pollingStarted.current` is only ever set to true when this hook is unmounted. Is that a problem?

Reply:
Er, I should have asked, is that by design?

Reply:
Yes, it is by design; it is a hacky work-around to deal with React Strict mode. When the page gets rendered twice in strict mode, two kiosk codes would get generated. With the issues that I've had with the timeouts, this generated a lot of errors because the backend would throttle and cut off the instance/device very quickly. 

I know that this is a sign that `useEffect` is not being used correctly, but I was unsure of a better solution because based on my current knowledge of `useEffect`, it was the hook that I was looking for. 

 I was thinking of splitting all of the polling logic into a separate component, but I wasn't sure doing that would actually solve a problem.

Reply:
Oh I see now! Separating the polling from the component would be a good idea. React components can re-render any number of times, makes managing linked async processes extra challenging.